allennlp/training/trainer.py

```diff
@@ -552,25 +559,25 @@ def _train_epoch(self, epoch: int) -> Dict[str, float]:
                 cuda_device=[self.cuda_device],
             )
 
-            # Updating tqdm only for the master as the trainers wouldn't have one
             if self._master:
+                # Updating tqdm only for the master as the trainers wouldn't have one
                 description = training_util.description_from_metrics(metrics)
                 batch_group_generator_tqdm.set_description(description, refresh=False)
                 self._tensorboard.log_batch(
                     self.model, self.optimizer, batch_grad_norm, metrics, batch_group, param_updates
                 )
 
-            if self._master:
                 self._checkpointer.maybe_save_checkpoint(self, epoch, batches_this_epoch)
-                for callback in self._batch_callbacks:
-                    callback(
-                        self,
-                        batch_group,
-                        batch_group_outputs,
-                        epoch,
-                        batches_this_epoch,
-                        is_training=True,
-                    )
+            for callback in self._batch_callbacks:
+                callback(
+                    self,
+                    batch_group,
+                    batch_group_outputs,
+                    epoch,
+                    batches_this_epoch,
+                    is_training=True,
+                    is_master=self._master,
```

Comment:
If the issue is the opaque private argument, I'd vote for adding a simple `is_master()` method to the trainer, instead of passing yet another flag here.

Reply:
I'll be easily swayed one way or another in this matter, but my reasoning for having it this way was this: When you implement a `BatchCallback`, it's easy to forget about the multi-process case, but you almost certainly need to think about it. By making it a parameter, it becomes more visible and harder to ignore.

Reply:
That's a good point that I hadn't thought of.  Given that, I could also go either way here.  Whatever you think is best.  Is the epoch callback only ever called from master?

Reply:
Yes, the epoch callback is only called from master. I thought it would make more sense, but open reflection, I'm not sure that's true. I'll make another PR that adds the same thing to the epoch callback.

Reply:
Actually, that wasn't even true. The epoch callback was called all the time. Glad I checked!